soc: nordic: Update UICR format and add secondary firmware support

SebastianBoe opened the pull request:
Update Nordic UICR format to latest revision (v2.0) and add secondary firmware support to gen_uicr.py script.

Changes:
- Update UICR C struct with new Secondary firmware fields
- Add secondary firmware configuration support in gen_uicr.py
- New command-line options for secondary firmware address and PERIPHCONF

These changes support Nordic's IronSide SE architecture migration.

SebastianBoe committed:
soc: nordic: Update UICR format

Update the C struct for UICR to the latest revision.

Signed-off-by: Sebastian Bøe <sebastian.boe@nordicsemi.no>

SebastianBoe committed:
soc: nordic: gen_uicr: Support secondary firmware

Add support for secondary firmware in gen_uicr.py.

Signed-off-by: Sebastian Bøe <sebastian.boe@nordicsemi.no>

fabiobaltieri commented:
Also feels like this could as well be three separate PRs, we don't normally allow a single PR to contain multiple unrelated changes, probably no big deal for this one as they are alll nordic specific but it's normally desirable to have changes looked separately, potentially by different maintainers etc...

SebastianBoe commented:
> Also feels like this could as well be three separate PRs, we don't normally allow a single PR to contain multiple unrelated changes, probably no big deal for this one as they are alll nordic specific but it's normally desirable to have changes looked separately, potentially by different maintainers etc...

@fabiobaltieri 
Created a separate PR for the bootmode service.
The other two commits have a dependency between them.
https://github.com/zephyrproject-rtos/zephyr/pull/95410

SebastianBoe commented:
> Please provide a meaningful PR title and description. Thanks! https://docs.zephyrproject.org/latest/contribute/contributor_expectations.html#pr-requirements

@kartben 

done

jonathannilsen commented on soc/nordic/common/uicr/gen_uicr/CMakeLists.txt:
```diff
@@ -62,7 +64,12 @@ if(CONFIG_GEN_UICR_GENERATE_PERIPHCONF)
 
     if(EXISTS ${_dir}/zephyr/zephyr.dts)
-      list(APPEND periphconf_elfs ${_dir}/zephyr/zephyr.elf)
+      # Check if this is secondary firmware by looking for the marker file
+      if(EXISTS ${_dir}/is_secondary_firmware.txt)
+        list(APPEND secondary_periphconf_elfs ${_dir}/zephyr/zephyr.elf)
+      else()
+        list(APPEND periphconf_elfs ${_dir}/zephyr/zephyr.elf)
+      endif()
```
Maybe there is a way to distinguish between primary/secondary at a lower level, even including that information in the ELF itself? For example, if we defined two different periphconf sections in the elf instead of just one, and chose which to populate based on some config in the image (changing where we put things with the macros in uicr.h). Then each input file would be self describing, and this part wouldn't need a marker file.

SebastianBoe replied:
I'm sure that is possible.

But it seems a lot more complicated than just having a text file that signifies if an image is a secondary image or not.

It is easier both for humans and the build system to create and read files than it is for humans and build systems to modify and read elf files.

SebastianBoe commented:
@fabiobaltieri : I'd like to contribute a sample that uses/tests the features implemented in (what is now two PRs):

 https://github.com/zephyrproject-rtos/zephyr/pull/95451
 and this PR.
 
 How would you recommend that I organize this?
 
a) Create a new PR that contains both PRs and a commit for the new test?
b) Close  https://github.com/zephyrproject-rtos/zephyr/pull/95451 and put everything in 95410 ?

jonathannilsen commented on soc/nordic/common/uicr/gen_uicr.py:
```diff
@@ -104,6 +119,64 @@ class Mpcconf(c.LittleEndianStructure):
+    ]
+
+
+class SecondaryProtectedmem(c.LittleEndianStructure):
+    _pack_ = 1
+    _fields_ = [
+        ("ENABLE", c.c_uint32),
+        ("SIZE4KB", c.c_uint32),
+    ]
+
+
+class SecondaryWdtstart(c.LittleEndianStructure):
+    _pack_ = 1
+    _fields_ = [
+        ("ENABLE", c.c_uint32),
+        ("INSTANCE", c.c_uint32),
+        ("CRV", c.c_uint32),
+    ]
+
+
+class SecondaryPeriphconf(c.LittleEndianStructure):
+    _pack_ = 1
+    _fields_ = [
+        ("ENABLE", c.c_uint32),
+        ("ADDRESS", c.c_uint32),
+        ("MAXCOUNT", c.c_uint32),
+    ]
+
+
+class SecondaryMpcconf(c.LittleEndianStructure):
+    _pack_ = 1
+    _fields_ = [
+        ("ENABLE", c.c_uint32),
+        ("ADDRESS", c.c_uint32),
+        ("MAXCOUNT", c.c_uint32),
+    ]
```
These should never differ from their non-secondary counterparts, so I think we can just reuse those structures instead of defining new ones

jonathannilsen commented on soc/nordic/common/uicr/gen_uicr.py:
```diff
@@ -66,23 +66,38 @@ class Protectedmem(c.LittleEndianStructure):
+
+
+class SecurestorageCrypto(c.LittleEndianStructure):
+    _pack_ = 1
+    _fields_ = [
+        ("APPLICATIONSIZE1KB", c.c_uint32),
+        ("RADIOCORESIZE1KB", c.c_uint32),
     ]
 
 
-class Its(c.LittleEndianStructure):
+class SecurestorageIts(c.LittleEndianStructure):
+    _pack_ = 1
+    _fields_ = [
+        ("APPLICATIONSIZE1KB", c.c_uint32),
+        ("RADIOCORESIZE1KB", c.c_uint32),
+    ]
```
These are also expected to always be identical to each others, so one class could be used for both

SebastianBoe replied:
see https://github.com/zephyrproject-rtos/zephyr/pull/95410#discussion_r2321532580

SebastianBoe commented:
@fabiobaltieri I have split up the PR as requested, but this is causing issues.

https://github.com/zephyrproject-rtos/zephyr/pull/95410#issuecomment-3252500902

SebastianBoe commented:
Could not reproduce CI failure. Rebasing in hope it goes away.

SebastianBoe committed:
soc: noric: nrf54h20: Fix custom CONFIG_KERNEL_BIN_NAME bug

Fix bug where users were unable to name their binary Bøe when building
for nrf54h20.

Signed-off-by: Sebastian Bøe <sebastian.boe@nordicsemi.no>
